### PR TITLE
[AUDIO] Fix audio UTC failure issues

### PR DESCRIPTION
**Pull request by kishore-sn**
Below issues fixed:
Missing underrun handling code in some UTC's
Fix I2S stop / pause / resume API's which was causing failure in MMAP related audio UTC's

### CODE REVIEW

**Comment by seinfra** [comment]
Code Rule Check Result:
os/arch/arm/src/s5j/s5j_i2s.c:1719: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/arch/arm/src/s5j/s5j_i2s.c:1719: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/s5j/s5j_i2s.c:1721: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/arch/arm/src/s5j/s5j_i2s.c:1721: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/s5j/s5j_i2s.c:1722: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/arch/arm/src/s5j/s5j_i2s.c:1722: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/arch/arm/src/s5j/s5j_i2s.c:1723: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/arch/arm/src/s5j/s5j_i2s.c:1723: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

**Comment by seinfra** [comment]
Code Rule Check OK.

File: os/arch/arm/src/s5j/s5j_i2s.c [hunk 1]
```diff
@@ -1600,18 +1600,20 @@ static int i2s_pause(struct i2s_dev_s *dev)
 {
 	struct s5j_i2s_s *priv = (struct s5j_i2s_s *)dev;
 	irqstate_t flags;
+	volatile u32 reg = 0;
 	DEBUGASSERT(priv);
 
+	reg = getreg32(priv->base + S5J_I2S_CON);
 #ifdef I2S_HAVE_TX_P
-	if (priv->txpenab) {
+	if (priv->txpenab && reg & I2S_CR_TXDMACTIVE) {
```
Comment (btheosam) [inline os/arch/arm/src/s5j/s5j_i2s.c]: Why this check of DMA active status? What happens if pause bit is set when dma is inactive? Any harm? Better you check the status within irqsave.

File: os/arch/arm/src/s5j/s5j_i2s.c [hunk 2]
```diff
@@ -1600,18 +1600,20 @@ static int i2s_pause(struct i2s_dev_s *dev)
...
 		flags = irqsave();
 		modifyreg32(priv->base + S5J_I2S_CON, 0, I2S_CR_TXCHPAUSE);
 		irqrestore(flags);
 	}
 #endif
 
 #ifdef I2S_HAVE_RX
-	if (priv->rxenab) {
+	if (priv->rxenab && reg & I2S_CR_RXDMACTIVE) {
```
Comment (btheosam) [inline os/arch/arm/src/s5j/s5j_i2s.c]: Same as prev

File: os/arch/arm/src/s5j/s5j_i2s.c [hunk 3]
```diff
@@ -1650,9 +1654,13 @@ static int i2s_stop(struct i2s_dev_s *dev)
 	struct s5j_i2s_s *priv = (struct s5j_i2s_s *)dev;
 	irqstate_t flags;
 	struct s5j_buffer_s *bfcontainer;
+	volatile u32 reg;
 	DEBUGASSERT(priv);
 
 #ifdef I2S_HAVE_TX_P
+	modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_TXDMACTIVE, 0);
```
Comment (btheosam) [inline os/arch/arm/src/s5j/s5j_i2s.c]: Why is this modified without disabling interrupts?

File: os/arch/arm/src/s5j/s5j_i2s.c [hunk 4]
```diff
@@ -1681,6 +1687,9 @@ static int i2s_stop(struct i2s_dev_s *dev)
 #endif
 
 #ifdef I2S_HAVE_RX
+	modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_RXDMACTIVE, 0);
```
Comment (btheosam) [inline os/arch/arm/src/s5j/s5j_i2s.c]: same as above

File: os/arch/arm/src/s5j/s5j_i2s.c (hunk 1 above)
Comment (kishore-sn) [reply os/arch/arm/src/s5j/s5j_i2s.c]: We need to pause DMA channel, only if it is currently active.
Not doing this was resulting in UTC getting blocked in pcm_wait() which waits for a dequeue msg.

I modified to use irqsave and perform the entire pause / resume as one single atomic operation.

File: os/arch/arm/src/s5j/s5j_i2s.c (hunk 2 above)
Comment (kishore-sn) [reply os/arch/arm/src/s5j/s5j_i2s.c]: Same as above

File: os/arch/arm/src/s5j/s5j_i2s.c (hunk 3 above)
Comment (kishore-sn) [reply os/arch/arm/src/s5j/s5j_i2s.c]: Changed

File: os/arch/arm/src/s5j/s5j_i2s.c (hunk 4 above)
Comment (kishore-sn) [reply os/arch/arm/src/s5j/s5j_i2s.c]: Changed.

**Comment by seinfra** [comment]
Code Rule Check OK.

File: os/arch/arm/src/s5j/s5j_i2s.c [hunk 5]
```diff
@@ -1600,47 +1600,47 @@ static int i2s_pause(struct i2s_dev_s *dev)
...
 #ifdef I2S_HAVE_RX
-	if (priv->rxenab) {
-		flags = irqsave();
+	if (priv->rxenab && reg & I2S_CR_RXDMACTIVE) {
 		modifyreg32(priv->base + S5J_I2S_CON, 0, I2S_CR_RXCHPAUSE);
-		irqrestore(flags);
 	}
 #endif
+	irqrestore(flags);
 	return 0;
```
Comment (girishks2000) [inline os/arch/arm/src/s5j/s5j_i2s.c]: you don't have to explicitly save and restore the irq state when you use  modifyreg32. It internally does it. take care of this in all places

File: os/arch/arm/src/s5j/s5j_i2s.c [hunk 6]
```diff
@@ -1650,9 +1650,14 @@ static int i2s_stop(struct i2s_dev_s *dev)
 	struct s5j_i2s_s *priv = (struct s5j_i2s_s *)dev;
 	irqstate_t flags;
 	struct s5j_buffer_s *bfcontainer;
+	volatile u32 reg;
 	DEBUGASSERT(priv);
```
Comment (girishks2000) [inline os/arch/arm/src/s5j/s5j_i2s.c]: added this but I cant see the usage of this anywhere

File: apps/examples/testcase/ta_tc/audio/utc/utc_audio_main.c [hunk 7]
```diff
@@ -728,6 +728,10 @@ static void utc_audio_pcm_drop_p(void)
 
 	while (remain > 0) {
 		frames_read = pcm_readi(g_pcm, buffer, pcm_get_buffer_size(g_pcm));
+		if (frames_read == -EPIPE) {
+			pcm_prepare(g_pcm);
+			continue;
+		}
 		if (frames_read < 0) {
```
Comment (girishks2000) [inline apps/examples/testcase/ta_tc/audio/utc/utc_audio_main.c]: Are you sure  it will succeed in reading / writing the second time.  There should be a way to check prepare status and try again.
Suggestion:
Retry mechanism has to be modified  based on some timeout mechanism. Or any better idea if you have

File: apps/examples/testcase/ta_tc/audio/utc/utc_audio_main.c [hunk 8]
```diff
@@ -861,7 +869,7 @@ static void utc_audio_pcm_writei_n(void)
 	ssize_t size;
 	char *buffer;
 
-	g_pcm = pcm_open(0, 0, PCM_IN, NULL);
+	g_pcm = pcm_open(0, 0, PCM_OUT, NULL);
 	TC_ASSERT_GT("pcm_writei", pcm_get_file_descriptor(g_pcm), 0);
```
Comment (girishks2000) [inline apps/examples/testcase/ta_tc/audio/utc/utc_audio_main.c]: Why direction changed? just curious. Was it a mistake earlier?

File: os/arch/arm/src/s5j/s5j_i2s.c [hunk 9]
```diff
@@ -1600,47 +1600,47 @@ static int i2s_pause(struct i2s_dev_s *dev)
...
+	volatile u32 reg = 0;
 	DEBUGASSERT(priv);
 
+	reg = getreg32(priv->base + S5J_I2S_CON);
+	flags = irqsave();
 #ifdef I2S_HAVE_TX_P
-	if (priv->txpenab) {
-		flags = irqsave();
+	if (priv->txpenab && reg & I2S_CR_TXDMACTIVE) {
 		modifyreg32(priv->base + S5J_I2S_CON, 0, I2S_CR_TXCHPAUSE);
```
Comment (girishks2000) [inline os/arch/arm/src/s5j/s5j_i2s.c]: separate conditions with brackets. Logical and conditional expression evaluation will be easy to understand

File: apps/examples/testcase/ta_tc/audio/utc/utc_audio_main.c (hunk 8 above)
Comment (kishore-sn) [reply apps/examples/testcase/ta_tc/audio/utc/utc_audio_main.c]: It was a mistake earlier. Fixed it now.

File: os/arch/arm/src/s5j/s5j_i2s.c (hunk 6 above)
Comment (kishore-sn) [reply os/arch/arm/src/s5j/s5j_i2s.c]: It is used in line num 1717.

File: os/arch/arm/src/s5j/s5j_i2s.c (hunk 9 above)
Comment (kishore-sn) [reply os/arch/arm/src/s5j/s5j_i2s.c]: ok done

File: os/arch/arm/src/s5j/s5j_i2s.c (hunk 5 above)
Comment (kishore-sn) [reply os/arch/arm/src/s5j/s5j_i2s.c]: I want both rx and tx channels to pause / resume in one shot. 
So, I think  its better to disable irq explicitly.. Please let me know your opinion.

Comment (girishks2000) [reply os/arch/arm/src/s5j/s5j_i2s.c]: But both features are mutually exclusive. In a single codec either one will be active at any given point of time. So oneshot enable disable doesn't make sense.

File: apps/examples/testcase/ta_tc/audio/utc/utc_audio_main.c (hunk 7 above)
Comment (kishore-sn) [reply apps/examples/testcase/ta_tc/audio/utc/utc_audio_main.c]: Fixed. 
Added retry mechanism

File: os/arch/arm/src/s5j/s5j_i2s.c (hunk 5 above)
Comment (kishore-sn) [reply os/arch/arm/src/s5j/s5j_i2s.c]: Yes true. But we had not planned for it. 
In that case, we will have to modify all our API's to handle tx and rx separately. 
I will do the changes and publish them.

Comment (kishore-sn) [reply os/arch/arm/src/s5j/s5j_i2s.c]: Fixed.

**Comment by kishore-sn** [comment]
I modified I2S API's to accept channel information and operate only on that channel. 
Please review again.

**Comment by seinfra** [comment]
Code Rule Check Result:
os/arch/arm/src/s5j/s5j_i2s.c:1683: ERROR: [BRC_M_SMT] that open brace { should be on the previous line

**Comment by seinfra** [comment]
Code Rule Check OK.

File: os/arch/arm/src/s5j/s5j_i2s.c [hunk 10]
```diff
@@ -1596,117 +1596,128 @@ static uint32_t i2s_txdatawidth(struct i2s_dev_s *dev, int bits)
...
-	s5j_dmastop(priv->txp.dma);
+	if (dir == I2S_TX) {
+		modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_FTXURINTEN, 0);
+		modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_TXDMACTIVE, 0);
+		s5j_dmastop(priv->txp.dma);
+
+		while (sq_peek(&priv->txp.pend) != NULL) {
+			flags = irqsave();
+			bfcontainer = (struct s5j_buffer_s *)sq_remfirst(&priv->txp.pend);
+			irqrestore(flags);
```
Comment (girishks2000) [inline os/arch/arm/src/s5j/s5j_i2s.c]: here also saving interrupt context is not required. You are working on a local copy of the data structure, so modification of when interrupt occurs should not affect. Same applies wherever local structure is modified

File: apps/examples/testcase/ta_tc/audio/utc/utc_audio_main.c [hunk 11]
```diff
@@ -810,6 +844,7 @@ static void utc_audio_pcm_writei_p(void)
 	char *buffer;
 	int num_read;
 	unsigned int size;
+	int retry = 10;
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/ta_tc/audio/utc/utc_audio_main.c]: #define PREPARE_RETRY_COUNT 10(Actually 10 is too large, I think )

File: os/arch/arm/src/s5j/s5j_i2s.c (hunk 10 above)
Comment (kishore-sn) [reply os/arch/arm/src/s5j/s5j_i2s.c]: As discussed, the queue is being updated in the irq context also. So we have to disable irq to protect the queue from corruption.

**Comment by Taejun-Kwon** [comment]
@GalkinIvan , could you please check s5j code here?

File: apps/examples/testcase/ta_tc/audio/utc/utc_audio_main.c (hunk 11 above)
Comment (kishore-sn) [reply apps/examples/testcase/ta_tc/audio/utc/utc_audio_main.c]: Done. 
I agree that retry count of 10 is big. But I think its ok for UTC. 
We can have a smaller count in media player.

**Comment by seinfra** [comment]
Code Rule Check OK.

File: os/arch/arm/src/s5j/s5j_i2s.c [hunk 12]
```diff
@@ -1596,117 +1596,128 @@ static uint32_t i2s_txdatawidth(struct i2s_dev_s *dev, int bits)
...
-	while (sq_peek(&priv->rx.done) != NULL) {
-		flags = irqsave();
-		bfcontainer = (struct s5j_buffer_s *)sq_remfirst(&priv->rx.done);
-		irqrestore(flags);
-		apb_free(bfcontainer->apb);
-		i2s_buf_rx_free(priv, bfcontainer);
+
+	reg = getreg32(priv->base + S5J_I2S_CON);
+
+	if ((reg & (I2S_CR_TXDMACTIVE | I2S_CR_RXDMACTIVE)) == 0) {
```
Comment (btheosam) [inline os/arch/arm/src/s5j/s5j_i2s.c]: This condition would disable I2S port when either of RX/TX is NOT active. But I see you want to handle RX & TX seperately. Confirm if this condition is okay.

**Comment by btheosam** [comment]
Not sure if this changes for I2S to add operations on RX/TX alone like i2s_stop(dev, RX ).
Coz, if ARTIK is the master, then to stop either RX/TX, the clocks had to be cut, which inturn would stop both RX & tX. Right now, alc5658 is the master, so stopping RX & TX at artik end might work, as the clocks still on from ALC end. Need a second opinion on this.  Unless these changes are generic and workable on both master & slave mode, this changes should be deferred.

File: os/arch/arm/src/s5j/s5j_i2s.c [hunk 13]
```diff
@@ -297,9 +297,9 @@ static uint32_t i2s_samplerate(struct i2s_dev_s *dev, uint32_t rate);
 
 static uint32_t i2s_txdatawidth(struct i2s_dev_s *dev, int bits);
 static int i2s_send(struct i2s_dev_s *dev, struct ap_buffer_s *apb, i2s_callback_t callback, void *arg, uint32_t timeout);
-static int i2s_stop(struct i2s_dev_s *dev);
-static int i2s_pause(struct i2s_dev_s *dev);
-static int i2s_resume(struct i2s_dev_s *dev);
+static int i2s_stop(struct i2s_dev_s *dev, i2s_ch_dir_t dir);
```
Comment (btheosam) [inline os/arch/arm/src/s5j/s5j_i2s.c]: Would it be possibel to stop only RX or TX if HOST is master?

File: os/arch/arm/src/s5j/s5j_i2s.c (hunk 12 above)
Comment (kishore-sn) [reply os/arch/arm/src/s5j/s5j_i2s.c]: It disables I2S when both RX and TX are not active.

File: os/arch/arm/src/s5j/s5j_i2s.c (hunk 13 above)
Comment (kishore-sn) [reply os/arch/arm/src/s5j/s5j_i2s.c]: Similar logic is applied in IRQ handler. 
So, I suppose this must work. 

Also, if stopping either RX or TX would completely stop the I2S clock, then why do we have separate register setting to stop the complete I2S core?

**Comment by GalkinIvan** [comment]
Changes of os/arch/arm/src/s5j/s5j_i2s.c look fine.

File: os/arch/arm/src/s5j/s5j_i2s.c (hunk 13 above)
Comment (btheosam) [reply os/arch/arm/src/s5j/s5j_i2s.c]: @kishore-sn Is it common for all chipset? (My knowledge is limited on other chipsets)
If you are sure that these changes are portable on Cypress board also, then I can dismiss the hold. 
But, if you are NOT sure, then better to wait for Ivan's agreement.

Comment (btheosam) [reply os/arch/arm/src/s5j/s5j_i2s.c]: @kishore-sn please dont scope only for ARTIK. Please check if this will work on others as well.

Comment (kishore-sn) [reply os/arch/arm/src/s5j/s5j_i2s.c]: I cannot be sure about how I2S works in all SOC's. In my opinion, there has to be a way to individually control the RX and TX channels. I have checked the I2S implementation in bcm43909 and also in an old android project based on Samsung SOC. They seem to have similar type of implementation. For example, checkout the below API's from bcm code:

static int i2s_stop_stream( platform_i2s_port_t port, platform_i2s_direction_t dir );

So, I suggest that we go ahead with this implementation for now. When we work on a new SOC, we would either need to find some way of implementing the similar feature or add a work around in our code. 

@Taejun-Kwon If you have any suggestions, please let me know.

Comment (Taejun-Kwon) [reply os/arch/arm/src/s5j/s5j_i2s.c]: I checked many i2s codes and I found seperate rx & tx with direction is common style.(I jsut checked android & zephyr's some h/w)
But I think benix's opinion also very important.
Let's discuss about it more. I wil invite you guys.
